pytorch_lightning/core/optimizer.py

```diff
@@ -162,3 +163,30 @@ def closure_dis():
         assert trainer is not None
         with trainer.profiler.profile(profiler_action):
             trainer.training_type_plugin.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
+
+
+class _MockOptimizer(Optimizer):
+    """The `_MockOptimizer` will be used inplace of an optimizer in the event that `None` is returned from
+    `configure_optimizers`."""
```

Comment:
(for a another issue/PR) @ethanwharris I saw this was added in #1279 - do you recall what the use case was? 

If users don't have optimizers, should they use manual optimization instead of returning None here?

this leaks out like this which could be very confusing for a user
```py
class MyLightningModule(LightningModule):
    def configure_optimizers():
        return None

    def training_step(batch, batch_idx):
        opt = self.optimizers()
        # opt is not None! what?!
```

Reply:
This goes back a bit haha. Not sure about the use case, someone wanted to disable the optimizer. If it's possible to use manual optimization to acheive this then I think it would be fine to remove it and document it somewhere.

Reply:
@ananthsub  @daniellepintz  Can we address this in a separate PR? We have use cases for None in configure_optimizers() and they depends on the configure_optimizers() could be None to achieve their implementations.

Reply:
@four4fish yes, I filed https://github.com/PyTorchLightning/pytorch-lightning/issues/11178 to discuss further.